File: PROG/Week04/Lab04/StretchWith2Vowels.py
import re


def read_sentence() -> list[str]:
    s = input("Sentence: ").lower()
    s = re.sub(r"\s+", " ", s)
    return s.split()


def is_vowel(c: str) -> bool:
    return c in "aeiou"


def count_vowels(segment: str) -> int:
    return sum([is_vowel(c) for c in segment])


def matches(word: str) -> bool:
    segments = word.split("z")
    return any([count_vowels(seg) == 2 for seg in segments])


def count_matches(sentence: list[str]) -> int:
    return sum([matches(word) for word in sentence])


if __name__ == "__main__":
    while True:
        s = read_sentence()
        if "*" in s:
            break
        m = count_matches(s)
        print(f"Matching words = {m}")
    print("Done")

File: PROG/Week04/Lab04/StretchWith2Vowels.java
public class StretchWith2Vowels {
    static String[] readSentence() {
        System.out.print("Sentence: ");
        String s = In.nextLine().toLowerCase();
        s = s.replaceAll("\\s+", " ");
        return s.split(" ");
    }

    static boolean Contains(String[] array, String target) {
        for (String string : array) {
            if (string.equals(target))
                return true;
        }
        return false;
    }

    static String[] vowels = { "a", "e", "i", "o", "u" };

    static boolean isVowel(String c) {
        return Contains(vowels, c);
    }

    static int countVowels(String segment) {
        String[] letters = segment.split("(?!^)");

        int sum = 0;
        for (String letter : letters) {
            sum += isVowel(letter) ? 1 : 0;
        }

        return sum;
    }

    static boolean matches(String word) {
        String[] segments = word.split("z");
        for (String segment : segments) {
            if (countVowels(segment) == 2)
                return true;
        }
        return false;
    }

    static int countMatches(String[] sentence) {
        int sum = 0;
        for (String word : sentence) {
            sum += matches(word) ? 1 : 0;
        }
        return sum;
    }

    public static void main(String[] args) {
        while (true) {
            String[] s = readSentence();
            if (Contains(s, "*"))
                break;
            int m = countMatches(s);
            System.out.println("Matching words = " + m);
        }
        System.out.println("Done");
    }
}